stable_baselines/common/distributions.py

```diff
@@ -346,9 +369,22 @@ def __init__(self, nvec, flat):
         :param flat: ([float]) the categorical logits input
         """
         self.flat = flat
+        with tf.variable_scope("input", reuse=False):
+            no_mask = tf.zeros_like(self.flat)
+            self._action_mask_ph = tf.placeholder_with_default(no_mask, shape=self.flat.shape, name="action_ph")
         self.categoricals = list(map(CategoricalProbabilityDistribution, tf.split(flat, nvec, axis=-1)))
         super(MultiCategoricalProbabilityDistribution, self).__init__()
 
+    def create_action_mask(self):
+        prev = 0
+        for categorical in self.categoricals:
+            categorical.slice_action_mask(self.action_mask_ph, prev)
+            prev = prev + categorical.logits.shape[-1]
+
```

Comment:
I still don't like to use the loop when distributions, when the action dimension to become more and more, the training and evaluation will slow down.

Python processing loops is very inefficient. 😢

Reply:
This code only gets executed once, so the slowdown is minimal, however I agree this is messy. I'll adopt part of your approach. I'll add you as a co-author for this commit.